portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -35,21 +38,50 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  static final int LOAD_SIZE = 5;
   private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-  private final ArrayList<Comment> comments = new ArrayList<>();
+  private FetchOptions fetchOptions;
+  private ArrayList<Comment> comments;
+  private String cursor;
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    this.comments = new ArrayList<>();
+    int numComments = Integer.parseInt(request.getParameter("numComments"));
+    String type = request.getParameter("type");
+
+    switch (type) {
+      case "load":
+        numComments = LOAD_SIZE;
+        // fall through
+      case "reload":
+        this.cursor = null;
+        fetchOptions = FetchOptions.Builder.withLimit(numComments);
+        break;
+      case "append":
+        fetchOptions = FetchOptions.Builder.withLimit(LOAD_SIZE);
+        break;
+      default:
+        this.cursor = null;
+        fetchOptions = FetchOptions.Builder.withLimit(LOAD_SIZE);
+    }
+     
+    if (this.cursor != null) {
+      fetchOptions.startCursor(Cursor.fromWebSafeString(this.cursor));
+    }
+
     Query query = new Query("Comment");
-    PreparedQuery results = datastore.prepare(query);
-    for (Entity e : results.asIterable()) {
+    PreparedQuery pq = datastore.prepare(query);
+    QueryResultList<Entity> results = pq.asQueryResultList(fetchOptions);
+    for (Entity e : results) {
       String location = (String) e.getProperty("location");
       String link = (String) e.getProperty("link");
       String description = (String) e.getProperty("description");
       long id = e.getKey().getId();
       Comment c = new Comment(location, link, description, String.valueOf(id));
       comments.add(c);
     }
+    this.cursor = results.getCursor().toWebSafeString();
```

Comment:
Is the conversion to/from a webSafeString necessary? Could you just store the Cursor itself?